python/dgl/heterograph.py

```diff
@@ -0,0 +1,1214 @@
+        -------
+        Tensor
+            The popped representation
+        """
+        pass
+
+    def set_e_repr(self, data, etype, edges=ALL, inplace=False):
+        """Set edge(s) representation of a single edge type.
+
+        `data` is a dictionary from the feature name to feature tensor. Each tensor
+        is of shape (B, D1, D2, ...), where B is the number of edges to be updated,
+        and (D1, D2, ...) be the shape of the edge representation tensor.
+
+        All update will be done out of place to work with autograd unless the
+        inplace flag is true.
+
+        Parameters
+        ----------
+        data : tensor or dict of tensor
+            Edge representation.
+        etype : tuple[str, str, str]
+            The edge type, characterized by a triplet of source type name,
+            destination type name, and edge type name.
+        edges : edges
+            Edges can be a pair of endpoint nodes (u, v), or a
+            tensor of edge ids. The default value is all the edges.
+        inplace : bool
+            If True, update will be done in place, but autograd will break.
+        """
+        pass
+
+    def get_e_repr(self, etype, edges=ALL):
+        """Get edge(s) representation.
+
+        Parameters
+        ----------
+        etype : tuple[str, str, str]
+            The edge type, characterized by a triplet of source type name,
+            destination type name, and edge type name.
+        edges : edges
+            Edges can be a pair of endpoint nodes (u, v), or a
+            tensor of edge ids. The default value is all the edges.
+
+        Returns
+        -------
+        dict
+            Representation dict
+        """
+        pass
+
+    def pop_e_repr(self, etype, key):
+        """Get and remove the specified edge repr of a single edge type.
+
+        Parameters
+        ----------
+        etype : tuple[str, str, str]
+            The edge type, characterized by a triplet of source type name,
+            destination type name, and edge type name.
+        key : str
+          The attribute name.
+
+        Returns
+        -------
+        Tensor
+            The popped representation
+        """
+        pass
+
+    def apply_nodes(self, func, v=ALL, inplace=False):
+        """Apply the function on the nodes with the same type to update their
+        features.
+
+        If None is provided for ``func``, nothing will happen.
+
+        Notes
+        -----
+        An error is raised if the graph contains multiple node types.  Use
+
+            g[ntype]
+
+        to select nodes with type ``ntype``.
+
+        Parameters
+        ----------
+        func : callable or None
+            Apply function on the nodes. The function should be
+            a :mod:`Node UDF <dgl.udf>`.
+        v : int, iterable of int, tensor, optional
+            The (type-specific) node (ids) on which to apply ``func``. The
+            default value is all the nodes (with the same type).
```

Comment:
v can't be string?

Reply:
Instead of `g.apply_nodes('nodetype', func)`, one does `g['nodetype'].apply_nodes(func)`

Reply:
i wonder if it's too expensive to always call `g['nodetype'].apply_nodes`. after all, it constructs a subgraph for just updating nodes.

Reply:
given yesterday's discussion, i think we agree to adopt the following syntax:
```python
g[‘user’, ‘item’].send_and_recv(‘purchase’, fn...)
g[‘user’, ‘item’, ‘purchase’].pull(‘item’, fn...)
g.apply_nodes(‘item’, fn...)
```

Reply:
Can we make `g['nodetype']` lazy?  Like, not actually constructing the subgraph unless the user wants to query the graph topology itself?

Besides, what if the users want to do something like `g['item'].pull([1, 3, 5])`?

Reply:
I think people can still do `pull([1, 3, 5])`. we can use the input type to determine the behavior.

Reply:
We can teach user to write `user_g=g['user']; user_g.apply_nodes(func)`. Another advantage of this is if `user_g` is written wrong(such as `user_gg`), it can be detected at compilation stage (or by smart IDE).

Reply:
> I think people can still do `pull([1, 3, 5])`. we can use the input type to determine the behavior.

No.  Users have to write
```python
g.pull('item', [1, 3, 5])
```
instead if we put node type inside the argument list.  The point is that the node list argument cannot be omitted in any case.

Reply:
what does a user do if he wants to pull on one type of nodes in a bipartite graph of 'user' and 'item'? following your syntax, i guess a user needs to do
```
g.pull('item', ALL, ...)
```

Reply:
> what does a user do if he wants to pull on one type of nodes in a bipartite graph of 'user' and 'item'? following your syntax, i guess a user needs to do
> 
> ```
> g.pull('item', ALL, ...)
> ```
> 
> if so, we still need to extend the API to accept a node type in pull.

In this case they need to write `g['user', 'item', 'purchase'].send_and_recv(...)` instead.

Reply:
so `pull` can't be used in a bipartite graph? there will be a performance issue. in general, `update_all` is much more efficient than `pull`, `pull` is much more efficient than `send_and_recv`. I think we should allow all of these API at a bipartite graph for the sake of efficiency.

also, what confuses me is that sometimes you say the following is supported.
```python
g.pull('item', [1, 3, 5])
```
sometimes, you say the API doesn't accept node type as input. i'm not sure what is supported and what is not supported.

Reply:
Makes sense.  I'll add that.